### PR TITLE
fix(python): handle non union self referencing schema dependencies

### PR DESCRIPTION
**Pull request by armandobelardo**
This PR introduces a number of changes to the codebase, including:

- The addition of a new version, `4.0.0-rc5`, with an `irVersion` of `53` and a `changelogEntry` of `fix`.
- The introduction of a new function, `get_non_union_types_with_self_referencing_dependencies`, which returns a set of type IDs for non-union types with self-referencing dependencies.
- The modification of the `TestCaseWithExpectedResult` class to use a string type for the `test_case` field, and the addition of a new field, `expected_result`, of type `VariableValue`.
- The addition of a new class, `TestCaseResultWithStdout`, which has a `result` field of type `TestCaseResult` and a `stdout` field of type `str`.
- The modification of the TraceResponse` class to use a string type for the `stack` field.
- The introduction of a new function, `update_forward_refs`, which updates forward references for a given class.

## Changes

- The `generators/python/sdk/versions.yml` file has been updated to include a new version, `4.0.0-rc5`, with an `irVersion` of `53` and a `changelogEntry` of `fix`.
- The `generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py` file has been modified to include a new function, `get_non_union_types_with_self_referencing_dependencies`, which returns a set of type IDs for non-union types with self-referencing dependencies.
- The `generators/python/src/fern_python/generators/pydantic_model/fern_aware_pydantic_model.py` file has been updated to use the new `get_non_union_types_with_self_referencing_dependencies` function in the `is_referencing_circular_reference` check.
- The `seed/python-sdk/trace/src/seed/commons/types/test_case_with_expected_result.py` file has been modified to use a string type for the `test_case` field in the `TestCaseWithExpectedResult` class, and to add a new field, `expected_result`, of type `VariableValue`.
- The `seed/python-sdk/trace/src/seed/problem/types/create_problem_request.py` file has been updated to use a string type for the `problem_description` field in the `CreateProblemRequest` class, and to add a new field, `input_params`, of type `typing.List["VariableTypeAndName"]`.
- The `seed/python-sdk/trace/src/seed/problem/types/problem_info.py` file has been modified to use a string type for the `problem_description` field in the `ProblemInfo` class, and to add a new field, `input_params`, of type `typing.List["VariableTypeAndName"]`.
- The `seed/python-sdk/trace/src/seed/submission/types/code_execution_update.py` file has been updated to use a string type for the `test_cases` field in the `CodeExecutionUpdate_Graded` class.
- The `seed/python-sdk/trace/src/seed/submission/types/graded_response.py` file has been modified to use a string type for the `test_cases` field in the `GradedResponse` class.
- The `seed/python-sdk/trace/src/seed/submission/types/stack_frame.py` file has been updated to use a string type for the `scopes` field in the `StackFrame` class.
- The `seed/python-sdk/trace/src/seed/submission/types/stack_information.py` file has been modified to use a string type for the `top_stack_frame` field in the `StackInformation` class.
- The `seed/python-sdk/trace/src/seed/submission/types/submission_status_for_test_case.py` file has been updated to use a string type for the `result` field in the `SubmissionStatusForTestCase_Graded` class, and to add a new field, `value`, of type `TestCaseGrade`.
- The `seed/python-sdk/trace/src/seed/submission/types/submission_type_state.py` file has been modified to use a string type for the `default_test_cases` and `custom_test_cases` fields in the `SubmissionTypeState_Test` class.
- The `seed/python-sdk/trace/src/seed/submission/types/test_case_result_with_stdout.py` file has been updated to use a string type for the `result` field in the `TestCaseResultWithStdout` class.
- The `seed/python-sdk/trace/src/seed/submission/types/test_submission_state.py` file has been modified to use a string type for the `default_test_cases` and `custom_test_cases` fields in the `TestSubmissionState` class.
- The `seed/python-sdk/trace/src/seed/submission/types/test_submission_status_v_2.py` file has been updated to use a string type for the `problem_info` field in the `TestSubmissionStatusV2` class.
- The `seed/python-sdk/trace/src/seed/submission/types/trace_response.py` file has been modified to use a string type for the `stack` field in the `TraceResponse` class.
- The `seed/python-sdk/trace/src/seed/submission/types/trace_response_v_2.py` file has been updated to use a string type for the `stack` field in the `TraceResponseV2` class.
- The `seed/python-sdk/trace/src/seed/submission/types/trace_responses_page.py` file has been modified to use a string type for the `trace_responses` field in the `TraceResponsesPage` class.
- The `seed/python-sdk/trace/src/seed/submission/types/trace_responses_page_v_2.py` file has been updated to use a string type for the `trace_responses` field in the `TraceResponsesPageV2` class.
- The <co: 20>`seed/python-sdk/trace/src/seed/submission/types/traced_test_case.py` file has been modified to use a string type for the `result` field in the `TracedTestCase` class.</co: 2

### CODE REVIEW

File: generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py [hunk 1]
```diff
@@ -53,6 +53,17 @@ def __init__(
             ):
                 self._non_union_self_referencing_type_ids.add(id)
 
+        self._non_union_types_with_self_referencing_dependencies: Dict[str, Set[str]] = dict()
```
Comment (dsinghvi) [inline generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py:56]: why do you need to make this a dict?

Comment (dsinghvi) [inline generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py:56]: # This map goes from every non union type to a list of referenced types that circularly reference themselves
nit: Dict[ir_types.TypeId, Set[ir_types.TypeId]]

Comment (armandobelardo) [reply generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py:56]: Discussed offline, but it's just making it a smaller lookup, vs taking the field then getting the dependent types then doing the same lookup

**Review by dsinghvi — APPROVED** [review]

File: generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py (hunk 1 above)
Comment (armandobelardo) [reply generators/python/src/fern_python/generators/context/pydantic_generator_context_impl.py:56]: Updated!

**Pull request merged**